sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/BadRecordException.scala

```diff
@@ -67,16 +67,23 @@ case class PartialResultArrayException(
   extends Exception(cause)
 
 /**
- * Exception thrown when the underlying parser meet a bad record and can't parse it.
+ * Exception thrown when the underlying parser meets a bad record and can't parse it. Used for
+ * control flow between wrapper and underlying parser without overhead of creating a full exception.
  * @param record a function to return the record that cause the parser to fail
  * @param partialResults a function that returns an row array, which is the partial results of
  *                      parsing this bad record.
- * @param cause the actual exception about why the record is bad and can't be parsed.
+ * @param cause the actual exception about why the record is bad and can't be parsed. It's better
+ *                      to use lightweight exception here (e.g. without stacktrace), and throw an
+ *                      adequate (user-facing) exception only in case it's necessary after
+ *                      unwrapping `BadRecordException`
  */
 case class BadRecordException(
     @transient record: () => UTF8String,
     @transient partialResults: () => Array[InternalRow] = () => Array.empty[InternalRow],
-    cause: Throwable) extends Exception(cause)
+    cause: Throwable) extends Exception(cause) {
```

Comment:
do we never throw `BadRecordException` directly? then its `cause` doesn't matter and we can make it take a function to generate the actual error.

Reply:
I wanted to make something like this, but the diff turned out to be quite large (too large for that performance optimization ticket). But I can propose a compromise - make 2 constructors (one with the value, and the other one with the factory function), and proceed with a larger refactoring later.

Reply:
Oh I found one place: https://github.com/apache/spark/blob/7c728b2c2d6cbe2b8310cd8cf8aa4c2528b489c8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala#L720
It throws `QueryExecutionErrors.malformedRecordsDetectedInRecordParsingError`, but with a `BadRecordException` cause, which has a real cause embedded inside. I wonder if we would break this place by omitting the cause...

Reply:
Yea, it's user-facing there: 
![image](https://github.com/apache/spark/assets/166514719/caa61ac9-7c6a-40b1-b166-34454890fb22)

Reply:
In that case I propose the following:
- Merge this PR with the origjnal implementation (with the `MalformedCSVRecordException`)
- Replace `BadRecordException` in `parseJson` with something more pretty
- Refactor all parsers to pass exception cause in `BadRecordException` lazily using factory function

Reply:
`parse_json` is a new function and not release yet, we can still change it. It looks wrong to throw `BadRecordException` as a user-facing error. cc @gene-db

Reply:
Yes, we can change `parse_json` exception behavior. What is the desired behavior?

cc @chenhao-db

Reply:
In `VariantExpressionEvalUtils.parseJson`, we should not wrap the error with `BadRecordException`
```
        throw QueryExecutionErrors.malformedRecordsDetectedInRecordParsingError(
          input.toString, BadRecordException(() => input, cause = e))
```
should be
```
        throw QueryExecutionErrors.malformedRecordsDetectedInRecordParsingError(
          input.toString, e)
```

Reply:
Make a PR right away, since it's a trivial change: https://github.com/apache/spark/pull/46428